pkg/endpoint/policy.go

```diff
@@ -562,50 +620,56 @@ func (e *Endpoint) Regenerate(owner Owner) <-chan bool {
 	return newReq.ExternalDone
 }
 
-// TriggerPolicyUpdates indicates that a policy change is likely to
+// TriggerPolicyUpdatesLocked indicates that a policy change is likely to
 // affect this endpoint. Will update all required endpoint configuration and
-// state to reflect new policy and regenerate programs if required.
+// state to reflect new policy.
 //
-// Returns true if policy was changed and endpoints needs to be rebuilt
-func (e *Endpoint) TriggerPolicyUpdates(owner Owner) (bool, error) {
-	e.Mutex.Lock()
-	defer e.Mutex.Unlock()
+// Returns true if policy was changed and the endpoint needs to be rebuilt
+func (e *Endpoint) TriggerPolicyUpdatesLocked(owner Owner, opts models.ConfigurationMap) (bool, error) {
 	if e.Consumable == nil {
 		return false, nil
 	}
 
-	changed, err := e.regeneratePolicy(owner)
+	changed, err := e.regeneratePolicy(owner, opts)
 	if err != nil {
-		return changed, fmt.Errorf("%s: %s", e.StringID(), err)
+		return false, fmt.Errorf("%s: %s", e.StringID(), err)
 	}
 
-	return changed, err
+	e.getLogger().Debugf("TriggerPolicyUpdatesLocked: changed: %d", changed)
+
+	return changed, nil
 }
 
+// SetIdentity resets endpoints policy identity to 'id'.
+// Caller triggers policy regeneration if needed.
+// Called with e.Mutex Locked
 func (e *Endpoint) SetIdentity(owner Owner, id *policy.Identity) {
-	repo := owner.GetPolicyRepository()
 	cache := policy.GetConsumableCache()
 
-	repo.Mutex.Lock()
-	defer repo.Mutex.Unlock()
-
 	if e.Consumable != nil {
 		if e.SecLabel != nil && id.ID == e.Consumable.ID {
+			// Even if the numeric identity is the same, the order in which the
+			// labels are represented may change.
 			e.SecLabel = id
 			e.Consumable.Mutex.Lock()
 			e.Consumable.Labels = id
 			e.Consumable.LabelArray = id.Labels.ToSlice()
 			e.Consumable.Mutex.Unlock()
 			return
 		}
+		// XXX: This removes the consumable from the cache even if other endpoints
```

Comment:
What's XXX?

Reply:
Changed "XXX" to "TODO". My intention was to get reviewer's attention and maybe think it over and give some feedback :-)